File: bmlib/context_processor/llm_processor.py
# bmlib — shared library for biomedical literature tools
# Copyright (C) 2024-2026 Dr Horst Herb
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""A context processor that extracts through an LLM.

Ready-made subclass of :class:`~bmlib.context_processor.IterativeContextProcessor`
for the common case: text chunks — optionally carrying a relevance score
from a semantic search — summarised by a model, batch by batch, and
consolidated until the answer fits one context.

Every model call goes through a :class:`~bmlib.agents.BaseAgent`, so token
accounting, retries, JSON repair and provider routing are the ones the rest
of bmlib uses.
"""

from __future__ import annotations

import logging
from typing import Any

from bmlib.agents import BaseAgent
from bmlib.context_processor.base import IterativeContextProcessor, ProgressCallback
from bmlib.context_processor.data_types import (
    ConsolidatedItem,
    ExtractionResult,
    ProcessingConfig,
)

logger = logging.getLogger(__name__)

#: A chunk of text with the relevance score a search assigned it.
ScoredChunk = tuple[str, float]

#: Confidence recorded when the model does not report one of its own.
DEFAULT_EXTRACTION_CONFIDENCE = 0.9

DEFAULT_EXTRACTION_PROMPT = """Extract the key information relevant to this query.

Query: {query}

Content:
{content}

INSTRUCTIONS:
- Focus on information directly addressing the query
- Preserve important details, facts, and evidence
- Be concise but do not lose relevant information
- Return the extracted information as plain text

Extracted Information:"""

DEFAULT_CONSOLIDATION_PROMPT = """Consolidate and synthesise these extracted passages.

Query: {query}

Previously Extracted Information:
{content}

INSTRUCTIONS:
- Merge overlapping or redundant information
- Preserve all unique relevant details
- Maintain logical organisation
- Return consolidated information as plain text

Consolidated Information:"""

STRUCTURED_EXTRACTION_PROMPT = """Extract the key information relevant to this query.

Query: {query}

Content:
{content}

INSTRUCTIONS:
- Focus on information directly addressing the query
- Preserve important details, facts, and evidence
- Assess your confidence in the extraction (0.0 to 1.0)

Response format (JSON only):
{
    "extracted_content": "The extracted information...",
    "confidence": 0.9,
    "key_findings": ["finding 1", "finding 2"]
}

Respond ONLY with valid JSON."""

_REQUIRED_PLACEHOLDERS = ("{query}", "{content}")


class LLMChunkProcessor(IterativeContextProcessor):
    """Summarise batches of text chunks with an LLM, consolidating recursively.

    Accepts plain strings or ``(text, score)`` tuples — the shape a semantic
    search returns. Chunks are rendered with their score so the model can
    weigh them, batched to fit, and summarised; if the summaries together
    still exceed one context they are summarised again, with a prompt that
    says so.

    Args:
        agent: The agent every model call runs through. It carries the
            model, the token accounting and the retry behaviour.
        extraction_prompt: Template used at level 0. Must contain
            ``{query}`` and ``{content}``. Defaults to
            :data:`DEFAULT_EXTRACTION_PROMPT`, or
            :data:`STRUCTURED_EXTRACTION_PROMPT` when *use_structured_output*.
        consolidation_prompt: Template used at every level above 0. Same
            requirement; defaults to :data:`DEFAULT_CONSOLIDATION_PROMPT`.
        config: Processing configuration.
        progress_callback: Called with each progress update.
        use_structured_output: Ask for JSON at level 0 and read the model's
            own confidence and key findings from it, through the agent's
            JSON repair and retry.
        temperature: Overrides the agent's temperature for these calls.
        max_tokens: Overrides the agent's token ceiling for these calls.

    Raises:
        ValueError: If either prompt template lacks a required placeholder.
    """

    def __init__(
        self,
        agent: BaseAgent,
        extraction_prompt: str | None = None,
        consolidation_prompt: str | None = None,
        config: ProcessingConfig | None = None,
        progress_callback: ProgressCallback | None = None,
        use_structured_output: bool = False,
        temperature: float | None = None,
        max_tokens: int | None = None,
    ) -> None:
        super().__init__(config=config, progress_callback=progress_callback)
        self.agent = agent
        self.use_structured_output = use_structured_output
        self.extraction_prompt = extraction_prompt or (
            STRUCTURED_EXTRACTION_PROMPT if use_structured_output else DEFAULT_EXTRACTION_PROMPT
        )
        self.consolidation_prompt = consolidation_prompt or DEFAULT_CONSOLIDATION_PROMPT
        self.temperature = temperature
        self.max_tokens = max_tokens

        self._validate_template(self.extraction_prompt, "extraction_prompt")
        self._validate_template(self.consolidation_prompt, "consolidation_prompt")

    @staticmethod
    def _validate_template(template: str, name: str) -> None:
        """Check a prompt template carries the placeholders that get filled.

        Args:
            template: The template to check.
            name: The parameter name, for the error message.

        Raises:
            ValueError: If a placeholder is missing.
        """
        for placeholder in _REQUIRED_PLACEHOLDERS:
            if placeholder not in template:
                raise ValueError(f"{name} must contain the {placeholder} placeholder")

    @staticmethod
    def _render(template: str, query: str, content: str) -> str:
        """Fill a prompt template.

        Substitution is by replacement, not :meth:`str.format`, so a
        template may contain literal braces — a JSON example, a regex, a
        LaTeX fragment — without doubling them.
        """
        return template.replace("{query}", query).replace("{content}", content)

    # --- Item rendering ---

    def format_item(self, item: Any, index: int) -> str:
        """Render a chunk, showing its search score when it has one.

        Args:
            item: A ``(text, score)`` tuple or a plain string.
            index: Position within the batch.

        Returns:
            The chunk's text under a one-line header.
        """
        if isinstance(item, tuple) and len(item) == 2:
            text, score = item
            if isinstance(text, str) and isinstance(score, (int, float)):
                return f"[Chunk {index + 1}, score {float(score):.2f}]\n{text}"

        if isinstance(item, str):
            return f"[Item {index + 1}]\n{item}"

        logger.warning("Rendering an item of unexpected type %s", type(item).__name__)
        return f"[Item {index + 1}]\n{item}"

    def format_consolidated_item(self, item: ConsolidatedItem, index: int) -> str:
        """Render a summary from the level below, naming the level it came from."""
        level = item.metadata.get("recursion_level", 0)
        return f"[Consolidated level {level}, item {index + 1}]\n{item.content}"

    def split_oversized_item(self, item: Any, max_chars: int, overlap: int = 0) -> list[Any]:
        """Split a chunk too long to fit, keeping its score on every piece.

        Args:
            item: The chunk to split.
            max_chars: Maximum characters of content per piece.
            overlap: Characters of overlap between pieces.

        Returns:
            Pieces of the same shape as the input.
        """
        if isinstance(item, tuple) and len(item) == 2:
            text, score = item
            if isinstance(text, str) and isinstance(score, (int, float)):
                return [(piece, score) for piece in self._split_string(text, max_chars, overlap)]

        return super().split_oversized_item(item, max_chars, overlap)

    # --- Extraction ---

    def extract_from_batch(
        self,
        batch_content: str,
        query: str,
        batch_metadata: dict[str, Any],
    ) -> ExtractionResult:
        """Summarise one batch with the model.

        Level 0 uses the extraction prompt; every level above uses the
        consolidation prompt, which asks the model to merge rather than
        extract.

        Args:
            batch_content: The batch's formatted content.
            query: The question guiding extraction.
            batch_metadata: Carries ``recursion_level`` and ``batch_index``.

        Returns:
            The model's answer, with its confidence when it reported one.

        Raises:
            RuntimeError: If the model call fails. The base class records it
                against the batch.
        """
        level = batch_metadata.get("recursion_level", 0)
        structured = self.use_structured_output and level == 0
        template = self.extraction_prompt if level == 0 else self.consolidation_prompt
        prompt = self._render(template, query, batch_content)
        label = f"context extraction (level {level}, batch {batch_metadata.get('batch_index', 0)})"

        try:
            if structured:
                content, confidence, findings = self._extract_structured(prompt, label)
            else:
                response = self.agent.chat(
                    [self.agent.user_msg(prompt)],
                    temperature=self.temperature,
                    max_tokens=self.max_tokens,
                )
                content = response.content.strip()
                confidence = DEFAULT_EXTRACTION_CONFIDENCE
                findings = []
        except Exception as exc:
            raise RuntimeError(f"LLM extraction failed: {exc}") from exc

        metadata = dict(batch_metadata)
        if findings:
            metadata["key_findings"] = findings

        logger.debug(
            "Extracted %d chars from batch %s at level %d",
            len(content),
            batch_metadata.get("batch_index", 0),
            level,
        )
        return ExtractionResult(content=content, metadata=metadata, confidence=confidence)

    def _extract_structured(self, prompt: str, label: str) -> tuple[str, float, list[Any]]:
        """Run a JSON-mode extraction and read the model's own confidence.

        Args:
            prompt: The rendered prompt.
            label: Retry context, so a failure names the batch it came from.

        Returns:
            ``(content, confidence, key_findings)``.
        """
        parsed = self.agent.chat_json(
            [self.agent.user_msg(prompt)],
            temperature=self.temperature,
            max_tokens=self.max_tokens,
            retry_context=label,
            require_dict=True,
        )
        content = str(parsed.get("extracted_content", "")).strip()
        try:
            confidence = float(parsed.get("confidence", DEFAULT_EXTRACTION_CONFIDENCE))
        except (TypeError, ValueError):
            logger.warning("Model reported an unusable confidence in %s; using the default", label)
            confidence = DEFAULT_EXTRACTION_CONFIDENCE
        # A model that reports 1.4 or -0.2 must not defeat
        # ``min_confidence_threshold`` or skew a weighted merge.
        confidence = min(1.0, max(0.0, confidence))
        findings = parsed.get("key_findings") or []
        return content, confidence, list(findings) if isinstance(findings, list) else []

File: rust/bmlib/src/context_processor/llm_processor.rs
// bmlib — shared library for biomedical literature tools
// Copyright (C) 2024-2026 Dr Horst Herb
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

//! The one part of `context_processor/` that talks to a model.
//!
//! The harness itself has **no LLM dependency** — that is why this is a separate
//! module and why the crate needs no async runtime for it. This file is the whole
//! of the coupling, and it takes the model as a trait so the harness stays
//! testable without one.

use crate::context_processor::base::split_string;
use crate::context_processor::data_types::{ConsolidatedItem, ExtractionResult};
use regex::Regex;
use serde_json::Value;
use std::collections::BTreeMap;
use std::sync::OnceLock;

/// The confidence recorded when the model reported none.
pub const DEFAULT_EXTRACTION_CONFIDENCE: f64 = 0.9;

/// The placeholders a prompt template must carry.
pub const REQUIRED_PLACEHOLDERS: &[&str] = &["{query}", "{content}"];

fn placeholder_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r"\{(query|content)\}").expect("a fixed pattern"))
}

/// Check a prompt template carries the placeholders that get filled.
///
/// # Errors
///
/// The missing placeholder and the parameter's name, so a caller with two
/// templates knows which one to fix.
pub fn validate_template(template: &str, name: &str) -> Result<(), String> {
    for placeholder in REQUIRED_PLACEHOLDERS {
        if !template.contains(placeholder) {
            return Err(format!("{name} must contain the {placeholder} placeholder"));
        }
    }
    Ok(())
}

/// Fill a prompt template.
///
/// **Substitution is by replacement, not by a format string**, so a template may
/// contain literal braces — a JSON example, a regex, a LaTeX fragment — without
/// doubling them.
///
/// **One pass, not two chained replacements**: a second pass runs over what the
/// first substituted, so a query containing the literal `{content}` would have the
/// whole batch spliced into it — doubling a prompt that was sized to fit exactly,
/// which is the context overflow this module exists to prevent.
#[must_use]
pub fn render_template(template: &str, query: &str, content: &str) -> String {
    placeholder_pattern()
        .replace_all(template, |caps: &regex::Captures<'_>| {
            match caps.get(1).map(|m| m.as_str()) {
                Some("query") => query.to_string(),
                Some("content") => content.to_string(),
                _ => String::new(),
            }
        })
        .into_owned()
}

/// Whether an item is a `(text, score)` pair.
///
/// **Python requires a `tuple`**, and a JSON array is not one — so the source's
/// predicate is `false` for every array, including `["text", 0.5]`. That is a
/// distinction JSON cannot carry, and the port has to choose a spelling. It uses
/// a **two-element array whose first member is a string and whose second is a
/// number**, which is the shape the Python *means*: a caller handing this port a
/// scored chunk could not produce a Python tuple anyway, and reading arrays as
/// unscored would silently drop every score.
///
/// The two rules the source states are kept: a JSON **boolean is rejected
/// although it is a number** (`("text", True)` is a caller's mistake, and
/// rendering it as `score 1.00` hides that), and a three-element array is not a
/// pair.
#[must_use]
pub fn is_scored_chunk(item: &Value) -> bool {
    match item {
        Value::Array(items) if items.len() == 2 => {
            items[0].is_string() && items[1].is_number() && !matches!(items[1], Value::Bool(_))
        }
        _ => false,
    }
}

/// Render a chunk, showing its search score when it has one.
#[must_use]
pub fn format_item(item: &Value, index: usize) -> String {
    if is_scored_chunk(item) {
        let text = item[0].as_str().unwrap_or_default();
        let score = item[1].as_f64().unwrap_or(0.0);
        return format!("[Chunk {}, score {score:.2}]\n{text}", index + 1);
    }
    let text = match item {
        Value::String(text) => text.clone(),
        other => other.to_string(),
    };
    // The `[Item N]` header is used for a plain string **and** for an
    // unexpected type: the source logs a warning for the second and renders it
    // the same way, so the header alone does not distinguish them.
    format!("[Item {}]\n{text}", index + 1)
}

/// Render a summary from the level below, naming the level it came from.
#[must_use]
pub fn format_consolidated_item(item: &ConsolidatedItem, index: usize) -> String {
    // **The level is printed as the metadata spells it.** The source's f-string
    // coerces, so an integer `3` and the string `"3"` both render as `3`, and a
    // non-numeric value renders as itself rather than falling back to 0 — which a
    // numeric read here would do.
    let level = match item.metadata.get("recursion_level") {
        Some(Value::String(text)) => text.clone(),
        Some(Value::Number(n)) => n.to_string(),
        Some(other) => other.to_string(),
        None => "0".to_string(),
    };
    format!(
        "[Consolidated level {level}, item {}]\n{}",
        index + 1,
        item.content
    )
}

/// Split a chunk too long to fit, **keeping its score on every piece**.
///
/// A scored chunk loses its score if it is merely passed through the base
/// splitter, and the score is what the `WEIGHTED` consolidation strategy sorts on.
#[must_use]
pub fn split_oversized_item(item: &Value, max_chars: usize, overlap: usize) -> Vec<Value> {
    if is_scored_chunk(item) {
        let text = item[0].as_str().unwrap_or_default();
        let score = item[1].clone();
        return split_string(text, max_chars, overlap)
            .into_iter()
            .map(|piece| Value::Array(vec![Value::String(piece), score.clone()]))
            .collect();
    }
    if let Value::String(text) = item {
        return split_string(text, max_chars, overlap)
            .into_iter()
            .map(Value::String)
            .collect();
    }
    vec![item.clone()]
}

/// What the model is asked, and how.
#[derive(Debug, Clone, PartialEq)]
pub struct PromptTemplates {
    /// The level-0 extraction prompt. Must carry both placeholders.
    pub extraction_prompt: String,
    /// The consolidation prompt for every level above. Must carry both.
    pub consolidation_prompt: String,
}

impl PromptTemplates {
    /// Validate both templates.
    ///
    /// # Errors
    ///
    /// The name of whichever template is short a placeholder.
    pub fn validate(&self) -> Result<(), String> {
        validate_template(&self.extraction_prompt, "extraction_prompt")?;
        validate_template(&self.consolidation_prompt, "consolidation_prompt")
    }
}

/// One batch's answer from the model.
#[derive(Debug, Clone, PartialEq)]
pub struct ModelAnswer {
    /// The extracted or summarised text.
    pub content: String,
    /// The model's own confidence, already clamped.
    pub confidence: f64,
    /// The findings it named, when it named any.
    pub findings: Vec<Value>,
}

impl ModelAnswer {
    /// A plain answer at the default confidence.
    #[must_use]
    pub fn plain(content: impl Into<String>) -> Self {
        ModelAnswer {
            content: content.into(),
            confidence: DEFAULT_EXTRACTION_CONFIDENCE,
            findings: Vec::new(),
        }
    }
}

/// How the processor reaches a model.
///
/// A trait rather than a concrete client, so the extraction rules are testable
/// without a network — and so this module does not depend on `llm/`'s transport.
pub trait ContextModel {
    /// A plain completion.
    ///
    /// # Errors
    ///
    /// Whatever the transport reports.
    fn complete(&self, prompt: &str, temperature: f64, max_tokens: i64) -> Result<String, String>;

    /// A JSON-mode completion, read as the structured extraction shape.
    ///
    /// # Errors
    ///
    /// Whatever the transport reports, or a body that is not an object.
    fn complete_json(
        &self,
        prompt: &str,
        temperature: f64,
        max_tokens: i64,
    ) -> Result<Value, String>;
}

/// Read the model's own confidence out of a structured answer.
///
/// **An unusable confidence falls back to the default rather than to zero**: a
/// model that reported `"high"` has not said it is unsure, and recording zero
/// would let `min_confidence_threshold` discard a good extraction for a type
/// error. The clamp is applied after, so a reported `1.5` becomes `1.0` rather
/// than being read as missing.
#[must_use]
pub fn read_confidence(parsed: &Value) -> f64 {
    let raw = parsed.get("confidence");
    let value = match raw {
        Some(Value::Number(n)) => n.as_f64(),
        Some(Value::String(s)) => s.trim().parse::<f64>().ok(),
        _ => None,
    };
    match value {
        Some(value) => value.clamp(0.0, 1.0),
        None => DEFAULT_EXTRACTION_CONFIDENCE,
    }
}

/// Read the findings list, dropping anything that is not a list.
#[must_use]
pub fn read_findings(parsed: &Value) -> Vec<Value> {
    match parsed.get("key_findings") {
        Some(Value::Array(items)) => items.clone(),
        // A truthy non-list is **not** coerced into a one-element list: the
        // source's `list(findings)` over a string would split it into
        // characters, which is a shape no caller asked for.
        _ => Vec::new(),
    }
}

/// One batch to summarise.
///
/// A struct rather than seven positional parameters: the two that matter most —
/// the content and the query — are both `&str`, so a transposed pair would compile
/// and silently ask the wrong question of the wrong text.
#[derive(Debug)]
pub struct BatchRequest<'a> {
    /// The prompts, already validated.
    pub templates: &'a PromptTemplates,
    /// The batch's formatted content.
    pub batch_content: &'a str,
    /// The question guiding extraction.
    pub query: &'a str,
    /// Carries `recursion_level` and `batch_index`.
    pub batch_metadata: &'a BTreeMap<String, Value>,
    /// Sampling temperature.
    pub temperature: f64,
    /// Output cap.
    pub max_tokens: i64,
    /// Whether a level-0 batch asks for the structured shape.
    pub use_structured_output: bool,
}

/// Summarise one batch with the model.
///
/// **Level 0 uses the extraction prompt; every level above uses the
/// consolidation prompt**, which asks the model to merge rather than extract.
///
/// # Errors
///
/// A transport or parse failure, so the harness can record it against the batch
/// rather than losing the batch's index.
pub fn extract_from_batch(
    model: &dyn ContextModel,
    request: &BatchRequest<'_>,
) -> Result<ExtractionResult, String> {
    let BatchRequest {
        templates,
        batch_content,
        query,
        batch_metadata,
        temperature,
        max_tokens,
        use_structured_output,
    } = *request;
    let level = batch_metadata
        .get("recursion_level")
        .and_then(Value::as_u64)
        .unwrap_or(0) as usize;
    let structured = use_structured_output && level == 0;
    let template = if level == 0 {
        &templates.extraction_prompt
    } else {
        &templates.consolidation_prompt
    };
    let prompt = render_template(template, query, batch_content);

    let answer = if structured {
        let parsed = model.complete_json(&prompt, temperature, max_tokens)?;
        let content = parsed
            .get("extracted_content")
            .map(|v| match v {
                Value::String(s) => s.clone(),
                other => other.to_string(),
            })
            .unwrap_or_default()
            .trim()
            .to_string();
        ModelAnswer {
            content,
            confidence: read_confidence(&parsed),
            findings: read_findings(&parsed),
        }
    } else {
        let content = model.complete(&prompt, temperature, max_tokens)?;
        ModelAnswer::plain(content.trim())
    };

    let mut metadata = batch_metadata.clone();
    if !answer.findings.is_empty() {
        metadata.insert("key_findings".to_string(), Value::Array(answer.findings));
    }

    let mut result = ExtractionResult::new(answer.content);
    result.metadata = metadata;
    result.confidence = answer.confidence;
    result.recursion_level = level;
    result.batch_index = batch_metadata
        .get("batch_index")
        .and_then(Value::as_u64)
        .map(|v| v as usize);
    Ok(result)
}
